### ISSUE TITLE
.NET OpenCV3 for UWP

### ISSUE THREAD

**Issue by JimSEOW** [cvwin/cvwin#3]
This GitHub is our best chance to .NET for OpenCV3 UWP.
I hope the proposed managed c++ wrapper is still moving forwards. 
Perhaps the [issues raised in OpenCVSharp](https://github.com/shimat/opencvsharp/issues/339#issuecomment-309263260) could be addressed with your help?

**Comment by harrywye** [comment]
@JimSEOW I don't want to give you a false hope. "C++ wrapper" will be a major undertaking, and it will likely never happen. Please check out the "UWP build" of OpenCV and see if it suits your needs. It runs on all three platforms win32, win64, and arm. The currently released builds (on nuget) are a bit outdated, but you can "easily" build your own using the information on this repo.

A lot has happened since I last worked on this (like the release of VS 2017 and changes in dot net frameworks/dot net core), and a minor tweaking may be needed. But, I expect it to be straightforward. An instruction can be found in another thread on this "issue" page with regards to how to build "uwp-compatible" win binaries. If you want to share your build on nuget gallery, you can find nuspec files under `/uwp` in this repo.

Please let me know if you need any further help.

**Comment by JimSEOW** [comment]
@harrywye Currently, there is ONLY EMGU that provide Commercial .NET UWP wrapper for OpenCV (arm, x86, x64). For me, this is a REAL joke that Microsoft is advocating 50% of Build 2017 talking about AI and yet, there is no UWP .NET for AI computer vision using openCV. 

Before Xamarin, as a C# developer, I always feel like 2nd class treatment by what is available in e.g. Java etc. Being treat like second class by Microsoft is too much in the areas of UWP computer vision and 3D (non Unity3D) UWP programming. 

Microsoft need to truly democratize .NET 3D, Machine learning and computer vision for UWP.

**Comment by harrywye** [comment]
@JimSEOW I absolutely agree with you. Although OpenCV is a bit of a "niche", it is a very important one. The fact that there is no native C# (or C++/CX) support for OpenCV is a major problem for us Windows developers, and it shows lack of commitment from Microsoft. I had been so frustrated, but I really gave up. I've now moved on. Windows (or, HoloLens) is not a primary focus for me.

BTW, I don't know if you are aware, but there is _a tectonic shift in computer vision._ CV has been moving for the last few years from algorithm based (e.g., OpenCV) to learning based (e.g., as used by all cloud-based CV services). I'm not sure if it's just a fad or if it is here to stay, but I suggest you look into this learning based computer vision. All major cloud services, AWS, Azure, and Google, provide limited CV services. I played around with them a little bit, and they work rather well for a certain well-defined set of problems (like face detection and image classification). (BTW, Microsoft's CV service is inferior to, say, that of Google, by a mile. Microsoft has been all talks, but has not delivered when it comes to "intelligent services". Quoting your word, Microsoft has been really a "joke" in terms of providing libraries/services like computer vision and speech recognition/natural language processing, etc. to their loyal developers.)

Anyways, the real question is whether learning-based CV is flexible enough for all computer vision problems. That is, will it be able to completely replace algorithm based CV in the long run? I don't know, and it's really too early to tell. Even if that is the case (deep learning has proven to work amazingly well in "difficult" problem domains such as CV/NLP for the last few years), it'll take many years. We still need OpenCV. 

Meanwhile, for some limited set of problems, learning-based CV works remarkably well. If your problem fits into this category (like image classification), then forget OpenCV on Windows. Just use a cloud service. (A good thing is, you don't have to be stuck with Microsoft's inferior technologies even if you _have to_ build your apps on Windows.)

**Comment by JimSEOW** [comment]
@harrywye We deal with REAL Human Face with Real Facial Expression and learning based will soon catch up to reveal if someone is sufficiently ill to justify an increase in insurance. 

For that reason, Learning Based need to be NON cloud Based TOO for data privacy. 

We need to be smart on how to partition between client and cloud Learning Based computer vision. As hardware is getting more powerful, we need to send markup image data and NOT the raw Image data to the cloud.  

THIS IS WHY we need this BLOODLY .NET UWP openCV $/%$%&$&%$%& Microsoft for not getting it!!

**Comment by harrywye** [comment]
@JimSEOW I agree with you. Cloud-based services have limitations (like privacy issues, as you point out). Unfortunately, implementing your own learning-based CV on local premise will be a rather difficult (and time-consuming) problem at this point (unless you are a machine learning expert). Another (major) limitation of learning-based CV is that it requires _a huge amount of training data_ (depending on your specific problems). I once saw a number they used to train their face detection service, and it was in the millions (although I don't remember the exact number). It's very unlikely that we will have that much data at our disposal for a particular CV problem we are trying to solve, in general.

So, I agree, We need OpenCV, for now, and for long time to come.
